### ISSUE TITLE
Mintty incompatible with Nerd-Fonts or Font-Awesome yet.

### ISSUE THREAD

**Issue by nowox** [mintty/mintty#638]
I tried to install Nerd-Fonts and Font-Awesome (typically: https://github.com/ryanoasis/nerd-fonts/blob/master/patched-fonts/SourceCodePro/Regular/complete/Sauce%20Code%20Pro%20Nerd%20Font%20Complete%20Windows%20Compatible.ttf) on my Windows, but mintty doesn't recognize it. 

I tried to remove all the font with different abnormal glyphs in Fontforge, but even if I simply open a compatible font in Fontforge and I save it as is, it become not compatible with mintty. 

Windows command.exe also doesn't support such fonts, but VisualStudio or any other IDE do. 

Would it be possible for mintty to support that kind of fonts?

**Comment by mintty** [comment]
Mintty supports all fonts and works with this font:
`mintty -o Font="SauceCodePro NF"`
However, the font chooser dialog lists only fonts that are known to the system as monospace font. You may check whether that is handled properly with Fontforge.

**Comment by nowox** [comment]
Woot! 

I put this in my `mintty` shortcut and it works: 

```
mintty.exe -o Font="SauceCodePro Nerd Font" -
```

I must note that I first tried with the wrong name `mintty -o Font="SauceCodePro NF"` and I did not get any error message.

**Comment by mintty** [comment]
For me, after installing the font you linked to, `mintty -o Font="SauceCodePro NF"` works. Maybe, since you work on fonts with Fontforge, you have versions installed under multiple names?

**Comment by nowox** [comment]
I exactly used the font I linked and it didn't work. However the other issue I have is the dimension of the icons that looks a bit squeezed: https://ibb.co/dkdCBF

It seems mintty consider the used font as monospace and resize all glyphs to fit in the allowed size. Would it be possible to look like this: https://github.com/ryanoasis/vim-devicons/wiki/screenshots/v0.8.x/overall-screenshot.png

**Comment by mintty** [comment]
Mintty expands certain characters if appropriate, by a factor of two, but it does not squeeze characters to cell width (although I'm considering to do that in order to conciliate #615).
In the screenshot, these icons seem to be displayed at double-cell width. Mintty could in theory do that and in fact there is such an experimental option already (https://github.com/mintty/mintty/wiki/CtrlSeqs#wide-characters) but there are two problems:
* By which criteria should characters be expanded, preferrably based on Unicode information?
* The result would not be consistent with locale-based width information which most software relies on!

**Comment by nowox** [comment]
After playing around for few days, I realized that `SauceCodePro` doesn't play well with italic, but it is not often the case: 

![a](https://cloud.githubusercontent.com/assets/1132849/23646852/181ead22-0313-11e7-8630-1a557987f3c0.png)

![b](https://cloud.githubusercontent.com/assets/1132849/23646858/1e568afc-0313-11e7-8d90-21ca79cfe364.png)

This is perhaps not related to Nerd Fonts at all.

Regarding your remark I don't really understand what would be the problem to extend a character a bit if it's larger than a standard monospace character even though it would mask the character low it. I think those who will use such icons have to put enough spaces after it: 

![drawing1](https://cloud.githubusercontent.com/assets/1132849/23647078/33abf7ec-0314-11e7-9c44-e8b1e5d38206.png)

So I think your second proposition is a good one.

**Comment by mintty** [comment]
The clipping of italics is a consequence of the nature of fixed-cell-width terminals. It's strange though that is does not happen always in your screenshot, and I don't reproduce it with "SauceCodePro NF". Please provide your config (esp font size, maybe shortened config file) and sample text.
Clipping within italic text could be avoided by rendering the text from right to left (perhaps slowing down a bit), maybe I'll check that.

About wide glyphs (like icons), if the request is not to reserve extra space to them (as would be needed for sequences of wider characters like in #553) or even to display everything at proportional width (#538, which very few terminals have ever tried to implement), it should be possible to allow a hangover for them.
But note that this would require that no output to the right neighbour cell is written; also, mintty would need to introduce a distinction between empty cells and space cells, which may turn out to be hard, don't know. Which other terminals support this?

**Comment by nowox** [comment]
As you suspected I was using another Nerd Font which is called "SauceCodePro Nerd Font". I replaced it with the link I provided and this time I don't have any issues with the italic, but the font looks quite ugly compared to the previous one. One other point is that the glyphs are not deformed anymore. 

![u](https://cloud.githubusercontent.com/assets/1132849/23655777/ce7b5e9e-0336-11e7-85f6-51c5a34a129d.png)

Unfortunately I forgot to backup what was in my `C:\Windows\Fonts` folder before removing every `SauceCodePro`. I have to see if I still have the buggy version at home.

**Comment by mintty** [comment]
I cannot reproduce the occasionally clipped "V". Please provide a test case.

**Comment by mintty** [comment]
I have to revise my previous statement about mintty not squeezing characters. It does in fact have such a mechanism, which was intended for single-width letters (according to the current locale metrics) of double-width glyphs (from certain East Asian fonts), e.g. "œ". This means the mechanism is desirable and useful and cannot be just dropped; as said before, reasonable criteria for different handling need to be identified.
Also, even changing the behaviour isn't quite as straight-forward; double-width icons are arbitrarily getting clipped when scrolling the screen. Some analysis is needed.

**Comment by mintty** [comment]
Please provide reproducible input on how to generate the screenshots from https://github.com/mintty/mintty/issues/638#issuecomment-283945247 and the last one (with the italic). Which terminal and programs were used? Make a screen log please (not screen shot).

**Comment by nowox** [comment]
```
$ cat ~/.minttyrc
BoldAsFont=yes
BackgroundColour=21,21,21
Transparency=off
Font=DejaVuSansMono NF
Locale=en_US
Charset=UTF-8
Term=xterm-256color
RightClickAction=paste
FontHeight=13
SwitchShortcuts=no
WindowShortcuts=yes
ZoomShortcuts=yes
AltFnShortcuts=yes
FontIsBold=no
BoldAsColour=yes
FontSmoothing=full
AllowBlinking=yes
CursorType=line
CursorBlinks=no
ClickTargetMod=alt
Scrollbar=none
Columns=104
Rows=58
PgUpDnScroll=yes

$ vi -u NONE
```

Then in `vim8`:

```
i"Hello World!"
<ESC>
Y100p
:syn on
:set ft=c
:highlight String cterm=italic
:set cursorcolumn
```

You should get this:

![pp](https://cloud.githubusercontent.com/assets/1132849/23751205/89638d22-04d0-11e7-82d6-e7078b64c893.png)

**Comment by mintty** [comment]
No, I don't, no italics following your instructions. And what's vim8?
Instead of instructions how to create this picture you could also provide a screen log (with mintty --log or `script` from the util-linux package).

**Comment by nowox** [comment]
Sorry about that `vim8` is simply `vim` version 8:

```
$ vi --version | head -n1
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Sep 26 2016 09:52:22)
```

And here is the log:

https://gist.github.com/nowox/c07ed5d1d94d0e62634156736fb1217d

With this I get this:

![snap000270](https://cloud.githubusercontent.com/assets/1132849/23756845/49d49470-04e5-11e7-8d69-9c43b5d2364c.png)

**Comment by mintty** [comment]
This log, when output in mintty with `cat`, does not render any italics.
Please provide a *reproducible* test case, i.e. a log file that, when simply output to the terminal, leaves the disputed text and artefacts on the screen.

**Comment by nowox** [comment]
I don't really understand. You said you don't get any italic in mintty. With 

```
mintty --version
mintty 2.0.3 (i686-pc-cygwin)
(C) 2015 Andy Koppe
License GPLv3+: GNU GPL version 3 or later
There is no warranty, to the extent permitted by law.
```

When I do:

```
echo -e '\e[3mThis text is italic'
```

I get an italic text on the output. I am afraid I cannot do better test case.

**Comment by mintty** [comment]
OK, I can cut the log file somewhat; actually something like your one-liner is sufficient, just not *that* one-liner. I can see that e.g. a "d" is clipped before a "!". I do not see the clipping of "H" or the block over "o" from your screenshot above.
For some reason, the string "World!" is rendered in two chunks, "World" and "!". This explains the clipping.
I'll try to find out why "!" isn't included in the chunk.

**Comment by mintty** [comment]
Released some improvements in 2.7.5, please check.
Full overhang enabling for italics would be more fiddly however. Not yet sure whether it's worth the effort.

**Comment by nowox** [comment]
Nice improvements, thanks for your time. It is good to mention that only monospace fonts are listed. 

Regarding italics, the latest version on the master branch doesn't solve the issue on vim with `set cursorcolumn` on and this font enabled: 

```
Font=DejaVuSansMono NF
```

This is not a real issue for me since I am not using this font. And again this is another issue out of topic. 

So thanks anyway.

**Comment by mintty** [comment]
> Full overhang enabling for italics would be more fiddly however. Not yet sure whether it's worth the effort.

In any case, it was a challenge. Now supported.

**Comment by mintty** [comment]
Released 2.7.6.

**Comment by mintty** [comment]
Released 2.8.0.

**Comment by mintty** [comment]
Released 3.1.5 with explicit character width attribute support.
